core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilder.java

```diff
@@ -450,8 +459,34 @@ public FilterFunctionBuilder rewrite(QueryRewriteContext context) throws IOExcep
         return this;
     }
 
-    public static FunctionScoreQueryBuilder fromXContent(Function<String, ScoreFunctionParser<?>> scoreFunctionLookup,
-            QueryParseContext parseContext) throws IOException {
+    /**
+     * Looks up {@linkplain ScoreFunctionParser}s based on name and context. Needs context to check the ParseFieldMatcher and report nice
+     * errors.
+     */
+    public static class ScoreFunctionsRegistry {
```

Comment:
I really think this should not be part of this query class. The fact that it's only used here doesn't mean that some other components will not be able to use score functions one day, which are also pluggable. The idea is: 1) plug custom function through SearchModyle 2) if you need to lookup a function, inject the registry. This last thing I think shouldn't have a dependency on the query. Sorry for making your life hard here, I hope you see what I mean.